### ISSUE TITLE
Add a DataStore Delete-by-Search-Criteria API Like JavaScript Has

### ISSUE THREAD

**Issue by jamesonwilliams** [aws-amplify/amplify-android#921]
In JavaScript, but not in iOS, it is possible to delete multiple items from the DataStore in a single call. The customer selects a model type, and provides a search filter. The search filter may be [one of these predicates](https://docs.amplify.aws/lib/datastore/data-access/q/platform/js#predicates), or may be a "match-all" filter.

### Proposed New Android API
```java
/**
 * Deletes any items of the given class that match the provided search criteria.
 * For example, the following code would delete all Post that have a status of "Draft":
 * <code>
 *     Amplify.DataStore.delete(Post.class, Post.STATUS.eq("Draft"),
 *         deletions -> {}, failure -> {}
 *     );
 * </code>
 * @param itemClass Class of item(s) to delete
 * @param searchFilter Delete items that match this filter, only.
 * @param onItemsDeleted Called when all matching items have been deleted, or if there
 *                       are no matching items that need to be deleted.
 * @param onDeletionFailure Called If an error is encountered while deleting items
 */
<T extends Model> void delete(
        @NonNull Class<T> itemClass,
        @NonNull QueryPredicate searchFilter,
        @NonNull Consumer<Iterator<DataStoreItemChange<T>>> onItemsDeleted,
        @NonNull Consumer<DataStoreException> onDeletionFailure
);
```

### Summary of Platform Support
|                                | *iOS* | *Android* | *JavaScript* |
|--------------------------------|-----|---------|------------|
| *Delete a Model (Unconditional)* | Y   | Y       | Y          |
| *Delete a Model (w/ Predicate)*  | Y   | Y       | Y          |
| *Delete a Model (by type & ID)*  | Y\*   | Y\*       | Y\*         |
| *Delete 0..n by search*          | **N**   | **N**       | Y          |

\* The use case of deleting a model by its type and ID is enabled on all three platforms. However, the mechanism is different across all three. iOS includes an explicit API which accepts a model type and an ID. JavaScript provides this functionality through its 0..n delete-by-search-filter API: the user may provide the model type, and a predicate that matches an ID. Android provides a `justId` factory on its models, so that you can do: `Amplify.DataStore.delete(Post.justId("...."), { }, { })`.

### Overview of JavaScript Delete APIs
The JavaScript version of the DataStore supports [three different approaches to delete content from the DataStore](https://github.com/aws-amplify/amplify-js/blob/main/packages/datastore/src/datastore/datastore.ts#L780-L792). It allows the user to delete single items, as well as multiple items of a given type.

1. Unconditionally delete a single item:
```javascript
await DataStore.delete(post)
```
2. _Conditionally_ delete a single item, only if it meets some criteria:
```javascript
await DataStore.delete(post, p => p.status("eq", PostStatus.DRAFT))
```
3. Delete any/all items that match a provided search filter:
```javascript
await DataStore.delete(/* big P */ Post, p => p.status("eq", PostStatus.DRAFT));
```
Note: `Predicates.ALL` may be provided, which means "delete all `Post`s."

### Overview of iOS Delete APIs
The iOS version of the DataStore supports [three approaches to delete single items from the DataStore](https://github.com/aws-amplify/amplify-ios/blob/main/Amplify/Categories/DataStore/DataStoreCategoryBehavior.swift#L29-L35). It does not support deleting multiple items of the same type.

1. Unconditionally delete a model instance:
```swift
Amplify.DataStore.delete(post) { ... }
```
2. Conditionally delete a model instance, if it meets some criteria:
```swift
Amplify.DataStore.delete(post, where: Post.keys.status == PostStatus.draft) { ... }
```
3. Unconditionally delete a model instance, identified by type and ID:
```swift
Amplify.DataStore.delete(Post.self, withId: "123") { ... }
```

**Comment by Amplifiyer** [comment]
iOS library's DataStorePlugin does [support](https://github.com/aws-amplify/amplify-ios/blob/62946da67e6b82429d896dc374c0ac952653e240/AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/AWSDataStorePlugin%2BDataStoreBaseBehavior.swift#L112-L114) 0..n delete-by-search-filter API but not exposed through the category. @drochetti can we confirm to what extent this is already supported by the library (looking at the code the storage adapter seems to support this use case)? And if not what will be required to bridge the gap?

**Comment by jamesonwilliams** [comment]
@Amplifiyer Cool, thanks for the link. Just a bit more work on Android. Neither the Android DataStore category nor the `AWSDataStorePlugin` supports this right now. We would have to:
1. Add the API on the category,
2. And on the plugin;
3. Rework the `LocalStorageAdapter` interface to support a 0...n result type;
4. Update [the delete implementation in `SQLiteStorageAdapter`](https://github.com/aws-amplify/amplify-android/blob/main/aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapter.java#L419-L501) to generalize [the predicate logic](https://github.com/aws-amplify/amplify-android/blob/main/aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapter.java#L446-L450), [relax this check](https://github.com/aws-amplify/amplify-android/blob/main/aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapter.java#L468-L473), and [return information about multiple changes](https://github.com/aws-amplify/amplify-android/blob/main/aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapter.java#L479-L488).

**Comment by TrekSoft** [comment]
My main objection to this feature request is that AppSync currently only supports delete by ID. So if we present an interface to the customer where they could easily delete, say, 10,000 records with a simple filter they would probably be shocked (and upset) if they found out that was making 10k separate network calls to achieve that. 

It's quite easy if customers want to consciously do such a thing to query on the filter and then loop through those results and make 10k delete calls locally (which will then lead to the 10k delete  calls on the network).

**Comment by raphkim** [comment]
Resolved with #1106
